metric.go

```diff
@@ -4,61 +4,104 @@ import (
 	"encoding/binary"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"sort"
 )
 
-var errInvalidIntervalzero = errors.New("interval cannot be 0")
 var errInvalidOrgIdzero = errors.New("org-id cannot be 0")
 var errInvalidEmptyName = errors.New("name cannot be empty")
-var errInvalidEmptyMetric = errors.New("metric cannot be empty")
 var errInvalidMtype = errors.New("invalid mtype")
 var errInvalidTagFormat = errors.New("invalid tag format")
 
-type PartitionedMetric interface {
-	Validate() error
-	SetId()
-	// return a []byte key comprised of the metric's OrgId
-	// accepts an input []byte to allow callers to re-use
-	// buffers to reduce memory allocations
-	KeyByOrgId([]byte) []byte
-	// return a []byte key comprised of the metric's Name
-	// accepts an input []byte to allow callers to re-use
-	// buffers to reduce memory allocations
-	KeyBySeries([]byte) []byte
-}
+var bPool = NewBufferPool()
 
 //go:generate msgp
+type Point struct {
+	Val float64
+	Ts  uint32
+}
+
+func (m Point) Timestamp() uint32 {
+	return m.Ts
+}
+
+func (m Point) Value() float64 {
+	return m.Val
+}
+
+type MetricPoint struct {
+	Id string
+	Point
+}
+
+func (m *MetricPoint) GetId() string {
+	return m.Id
+}
 
 // MetricData contains all metric metadata (some as fields, some as tags) and a datapoint
 type MetricData struct {
-	Id       string   `json:"id"`
-	OrgId    int      `json:"org_id"`
-	Name     string   `json:"name"`
-	Metric   string   `json:"metric"`
-	Interval int      `json:"interval"`
-	Value    float64  `json:"value"`
-	Unit     string   `json:"unit"`
-	Time     int64    `json:"time"`
-	Mtype    string   `json:"mtype"`
-	Tags     []string `json:"tags" elastic:"type:string,index:not_analyzed"`
+	MetricMetadata
+	Point
+}
+
+// can be used by some encoders, such as msgp
+type MetricDataArray []*MetricData
+
+type MetricMetadata struct {
+	Id       string `json:"id"`
+	OrgId    int    `json:"org_id"`
+	Name     string `json:"name"`     // graphite format
+	Interval int    `json:"interval"` // minimum 10
+	Unit     string `json:"unit"`
+	Mtype    string `json:"mtype"`
+
+	// some users of MetricMetadata (f.e. MetricTank) might add a "name" tag
+	// to this slice which allows querying by name as a tag. this special tag
+	// should not be stored or transmitted over the network, otherwise it may
+	// just get overwritten by the receiver.
+	Tags []string `json:"tags"`
+}
+
+func (m *MetricMetadata) Metadata() *MetricMetadata {
+	return m
+}
+
+func (m *MetricMetadata) GetId() string {
+	return m.Id
+}
+
+// returns a id (hash key) in the format OrgId.md5Sum
+// the md5sum is a hash of the the concatination of the
+// metric + each tag key:value pair (in metrics2.0 sense, so also fields), sorted alphabetically.
+func (m *MetricMetadata) SetId() {
+	b := bPool.Get()
+	sort.Strings(m.Tags)
+	buffer := bytes.NewBuffer(b)
+	buffer.WriteString(m.Name)
+	buffer.WriteByte(0)
+	buffer.WriteString(m.Unit)
+	buffer.WriteByte(0)
+	buffer.WriteString(m.Mtype)
+	buffer.WriteByte(0)
```

Comment:
I know the old version before this patch did the same, but anyway: Isn't that one `buffer.WriteByte(0)` too many? If there are no tags then the last char of `buffer` is `0`, if there are `>0` then there's a double-`0` before the first tag.